### PR TITLE
JS-2310 Fix S1848 DOM-derived constructor arguments

### PR DESCRIPTION
**Pull request by francois-mora-sonarsource**
## Summary

- Follow trusted DOM-selection provenance through chained member/call expressions and single-write local bindings.
- Keep S1848 reporting for reassigned values and locally shadowed `document`, `$`, and `jQuery` selector-like APIs.

## Scope

Jira: JS-2310
Rule keys: `javascript:S1848`, `typescript:S1848`

## Peach baselines

- `JS-2310-javascript-S1848-peach-before-20260825T155949Z.csv` (Jira attachment 304028)
- `JS-2310-typescript-S1848-peach-before-20260825T155949Z.csv` (Jira attachment 304029)

## Validation

- `npx tsx --tsconfig packages/tsconfig.test.json --test packages/analysis/src/jsts/rules/S1848/unit.test.ts`
- `npm run bbf`
- `npm run bridge:test`

### CODE REVIEW

File: packages/analysis/src/jsts/rules/S1848/rule.ts [hunk 1]
```diff
@@ -374,49 +399,44 @@ function containsDomSelection(node: estree.Node, scope: Scope.Scope): boolean {
...
+  const writeExpr = getUniqueWriteReferenceBefore(variable, node);
+  if (!writeExpr) {
+    return false;
   }
-  return node;
+
+  // Follow a local initializer only. This permits DOM-derived member/call chains
+  // without performing interprocedural analysis.
+  return containsDomSelection(unwrapTypeAssertion(writeExpr), scope, visitedVariables);
 }
```
[the remark this answers is not in the document]
Comment (francois-mora-sonarsource) [reply packages/analysis/src/jsts/rules/S1848/rule.ts:456]: Declined: a cross-function read is intentionally not treated as a safe provenance proof. Even for `const`, the nested function can be called before the initializer runs (for example through module initialization or another callback), so suppressing S1848 here risks hiding a real issue. This rule is deliberately report-first when execution order cannot be established.

**Review by oleksandr-selehenenko-sonarsource — CHANGES_REQUESTED** [review]
One point needs handling before merge:

- rule.ts, isVariableFromDomSelection, the cross-function-scope guard: bailing out when the read is in a different function than the write is correct for a reassignable let/var, but wrong for a const. A const DOM element is provably DOM-derived regardless of when the closure runs, and this was suppressed before the change. Now it reports. Verified against master: master raises no issue, the branch raises S1848 on this input:
```
const chart = document.querySelector('#chart');
document.getElementById('btn').addEventListener('click', () => {
new Chart(chart);
});
```

This is a common pattern (select once at module scope, use in a handler), so it's a real FP regression. Suggestion: allow the cross-function read when the binding is single-assignment (const, or one write total) and just folhe guard only for
multi-write variables. Add a positive test for the const-in-closure case.

**Non-blocking:**

1. helpers.ts, unwrapTypeAssertion: it re-implements the shared unwrapTypeScriptExpression (helpers/ast.ts) but drops TSSatisfiesEx

```
const el = document.querySelector('#nt;
new Chart(el.getContext('2d'));
```
Reuse the shared helper and unwrap ChainExpression separately (a chain expression is not a type assertion, so folding it in here istisfies` test.

2. trusted-dom-bindings.ts: the narrowing is, but it also reports a few legitimate forms: `const $ = jQuery`, `declare const $: JQueryStatic`, and `const { document } = window`. Probably all. Worth adding tests that pin the intended narrowing so it doesn't drift later.

3. helpers.ts, getUniqueWriteReferenceBefore: easy to confuse with  etUniqueWriteReference in helpers/ast.ts. A one-line note on why ts needed would help. getNearestFunctionScope / getUniqueWriteReferenceBefore are also generic scope utilities - if they get reused, helpers/ fits better thes/ folder.

**Comment by francois-mora-sonarsource** [comment]
@oleksandr-selehenenko-sonarsource thanks for the review — addressed all points:

**Blocking:** fixed in 7a2957509 (and refined in fb3936c19). The cross-function-scope guard now only applies its strict equality check to reassignable bindings; a single-write binding (`const`, or a `let`/`var` written exactly once) is trusted as long as the read stays within the write's function scope (an ancestor/descendant check, not equality), and the write must still textually precede the read. Added a positive test for the exact `const chart` + `addEventListener` case from your report, plus tests for deeper nesting and for a write-scope-shadowing edge case Gitar caught along the way.

**Non-blocking, all addressed in fb3936c19:**
1. `unwrapTypeAssertion` now delegates TS-assertion unwrapping to the shared `unwrapTypeScriptExpression` (so `TSSatisfiesExpression` is covered too) and only handles `ChainExpression` unwrapping locally, alternating between the two since they can nest in either order.
2. Extended `isTrustedJQueryBinding`/`isTrustedDocumentBinding` to trust `const $ = jQuery`, `const { document } = window`, and ambient type-only declarations (`declare const $: JQueryStatic`, `declare const document: Document`), with pinning tests for each.
3. Added a doc comment on `getUniqueWriteReferenceBefore` distinguishing it from `getUniqueWriteReference` in `helpers/ast.ts`. Didn't relocate it to the shared `helpers/` folder since it isn't reused outside S1848 yet — happy to move it if/when that changes.

CI and Gitar are both green on the latest push. One thing — your review text seems to have gotten garbled in a couple of spots (e.g. "folhe", "'#nt", "isTisTypeAssertion") — I did my best to infer intent from context, but flag in case something got cut off and I missed a point.

**Review by oleksandr-selehenenko-sonarsource — APPROVED** [review]
lgtm
